Kudu.Core/SourceControl/Git/LibGit2SharpRepository.cs

```diff
@@ -0,0 +1,396 @@
+            {
+                return Path.Combine(RepositoryPath, ".git", "hooks", "git-credential-invalid.sh");
+            }
+        }
+
+        public ChangeSet GetChangeSet(string id)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                LibGit2Sharp.Commit commit = repo.Lookup<LibGit2Sharp.Commit>(id);
+
+                if (commit == null)
+                {
+                    return null;
+                }
+
+                return new ChangeSet(commit.Sha, commit.Author.Name, commit.Author.Email, commit.Message, commit.Author.When);
+            }
+        }
+
+        public void AddFile(string path)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                repo.Stage(path);
+            }
+        }
+
+        public bool Commit(string message, string authorName, string emailAddress)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                try
+                {
+                    repo.Stage("*");
+                    if (string.IsNullOrEmpty(authorName) ||
+                        string.IsNullOrEmpty(emailAddress))
+                    {
+                        repo.Commit(message);
+                    }
+                    else
+                    {
+                        repo.Commit(message, new Signature(authorName, emailAddress, DateTimeOffset.UtcNow));
+                    }
+                    return true;
+                }
+                catch(LibGit2Sharp.EmptyCommitException)
+                {
+                    return false;
+                }
+            }
+        }
+
+        public void Update(string id)
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                repo.Checkout(id, new CheckoutOptions { CheckoutModifiers = CheckoutModifiers.Force });
+            }
+        }
+
+        public void Update()
+        {
+            Update("master");
+        }
+
+        public void Push()
+        {
+            using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+            {
+                var masterBranch = repo.Branches["master"];
+                repo.Network.Push(masterBranch);
+            }
+        }
+
+        public void FetchWithoutConflict(string remoteUrl, string branchName)
+        {
+            ITracer tracer = _tracerFactory.GetTracer();
+            try
+            {
+                using (var repo = new LibGit2Sharp.Repository(RepositoryPath))
+                {
+                    var trackedBranchName = string.Format("{0}/{1}", _remoteAlias, branchName);
+                    var refSpec = string.Format("+refs/heads/{0}:refs/remotes/{1}", branchName, trackedBranchName);
+
+                    // Configure the remote
+                    var remote = repo.Network.Remotes.Add(_remoteAlias, remoteUrl, refSpec);
```

Comment:
we want to call remotes.delete only if the remotes.add succeeds.  otherwise, remote.delete could throw or, code wise, not matching remotes.add.   imagine, ...
1. anything throws before remotes.add call 
2. remotes.add itself failed